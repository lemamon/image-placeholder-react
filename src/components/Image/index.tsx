import { useEffect, useState } from "react";
import defaultPlaceholder from "./../../assets/images/placeholder.jpg";
import Container from "../Container";
import { IImage } from "../../types";

const Image = ({
  src,
  alt,
  placeholder = defaultPlaceholder,
  ...props
}: IImage) => {
  const [imageIsReady, setImageIsReady] = useState(false);

  useEffect(() => {
    const img = new (window as any).Image();
    img.src = src;
    img.onload = () => setImageIsReady(true);
  }, [src]);

  return (
    <Container
      data-testid="component-container"
      isReady={imageIsReady}
      bgImage={placeholder}
    >
      {imageIsReady ? (
        <img data-testid="component-image" src={src} alt={alt} {...props} />
      ) : (
        ""
      )}
    </Container>
  );
};

export default Image;
