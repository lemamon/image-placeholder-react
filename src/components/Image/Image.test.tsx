import { screen, render, waitFor } from "@testing-library/react";
import Image from ".";
import placeholder from "./../../assets/images/placeholder.jpg";

describe("Image component", () => {
  it("mount render placeholder", () => {
    render(
      <div style={{ width: "500px", height: "500px" }}>
        <Image
          src={`https://via.placeholder.com/500`}
          placeholder={placeholder}
        />
      </div>
    );

    const element = screen.getByTestId("component-container");
    expect(element).toHaveStyle(`background: url(${placeholder})`);
  });

  it("Not show placeholder when parent size is not defined", () => {
    render(
      <Image
        src={`https://via.placeholder.com/500`}
        placeholder={placeholder}
      />
    );

    const element = screen.getByTestId("component-container");
    expect(element.clientHeight).toBe(0);
  });

  it("The snapshot of the component  must be the same", () => {
    const { container } = render(
      <Image
        src={`https://via.placeholder.com/500`}
        placeholder={placeholder}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Show loaded image correct", async () => {
    render(
      <Image
        src={`https://via.placeholder.com/500`}
        placeholder={placeholder}
      />
    );

    await waitFor(() =>
      expect(
        screen.getByTestId("component-image", {}, { timeout: 500 })
      ).toBeInTheDocument()
    );
  });

  it("remove placeholder when image is loaded", () => {});
});
