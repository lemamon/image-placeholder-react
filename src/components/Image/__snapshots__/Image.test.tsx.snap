// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component The snapshot of the component  must be the same 1`] = `
<div
  class="sc-bdnxRM tHtPp"
  data-testid="component-container"
/>
`;
